Combine ibus serial rx and telemetry

unitware opened the pull request:
It's now possible to combine the serial rx and ibus telemetry on the sam fc uart on a single pin.

I'm combining the ibus ports of the FS-iA6B receiver with a diode and a resistor.

```
+---------+
| FS-iA6B |
|         |
| Ser RX  |---|<---\       +------------+
|         |        |       | FC         |
| Sensor  |---[R]--*-------| SerialTX   |
+---------+                +------------+
```

R = 10Kohm, Diode 1N4148 or similar.

![2016-12-19 22 27 41](https://cloud.githubusercontent.com/assets/6065378/21365422/98982d5c-c6f5-11e6-9e36-3a51d7ca699b.jpg)

Both uart tx and rx channels are used so it's not possible to use the spare pin for rx of something else.

Discussion and how to connect here: https://github.com/cleanflight/cleanflight/issues/2545

unitware commented:
It flies quite well in acro mode.

In angle and horizon i think it's a bit "wobbly but non crashing". Probably some filtering or PID settings that needs to be tweaked. I'd love some feedback on that - blackbox logs: https://www.dropbox.com/s/kooe0h3zqqe11o3/blackbox_log_2016-12-29_105630.TXT?dl=0

I tested angle mode, then horizon and finally acro.

ledvinap commented on src/main/rx/ibus.c:
```diff
@@ -70,61 +75,92 @@ bool ibusInit(const rxConfig_t *rxConfig, rxRuntimeConfig_t *rxRuntimeConfig)
 
-    if (ibusFramePosition == 0 && c != IBUS_SYNCBYTE)
-        return;
+    if (ibusFramePosition == 0 && c <= IBUS_BUFFSIZE) {
```
Receiving `0` or `c > IBUS_BUFFSIZE`  here will have catastrophic consequences..

unitware replied:
ouch, thanks.

ledvinap commented on src/main/rx/ibus.c:
```diff
@@ -70,61 +75,92 @@ bool ibusInit(const rxConfig_t *rxConfig, rxRuntimeConfig_t *rxRuntimeConfig)
 
     chksum = 0xFFFF;
-    for (i = 0; i < 30; i++)
+    for (i = 0; i < (ibusLength - 2); i++)
```
There is `calculateChecksum` function

unitware replied:
yes, thanks. I forgot to change that after the refactoring.

unitware commented:
flight tested again after code change, still flying :)

McGiverGim commented:
Hi! I'm waiting for pieces of my new quadcopter. In a Naze32 10DOF my idea is to use OSD (UART1), IBus (UART2), SoftSerial 1 (IBus telemetry) and SoftSerial 2 (GPS). If I do it in this way I can't connect a LED strip.

I understand that this mod will let me use SoftSerial 1 for LED Strip and change the IBus telemetry to the UART1 too. Is that correct?

If I want to test it, what must I do? I'm a newbie and I'm not sure of how to test it.

unitware commented:
That is correct + I would not run telemetry on soft uart at all since it is on 115200 baud. 

Follow the instructions in the documentation (in this pull request) and have a look in the picture on how to do the wiring. 

Skickat från min iPhone

> 22 jan. 2017 kl. 10:55 skrev McGiverGim <notifications@github.com>:
> 
> Hi! I'm waiting for pieces of my new quadcopter. In a Naze32 10DOF my idea is to use OSD (UART1), IBus (UART2), SoftSerial 1 (IBus telemetry) and SoftSerial 2 (GPS). If I do it in this way I can't connect a LED strip.
> 
> I understand that this mod will let me use SoftSerial 1 for LED Strip and change the IBus telemetry to the UART1 too. Is that correct?
> 
> If I want to test it, what must I do? I'm a newbie and I'm not sure of how to test it.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

McGiverGim commented:
My pieces will arrive in one out two weeks. I will try it. If I have some problem I'll let you know.

Thanks for your hard work!

hydra commented:
you seem to have some additional commits that are not needed, can you remove them from the PR so that the commit list only shows the commits to be added by this PR and so the 'Files changed' tab shows only the code changes needed for the PR?    No new PR needed, just force push over your old PR branch on your repo and this PR will update.  If it helps, create a new (backup) branch from current PR branch before you start.

McGiverGim commented:
Hi! My Naze32 has arrived. Do you have a .hex with this modification to flash it? Or must I generate it by my own? I'm newbie with GIT , so I will need some instructions to accomplish that.

Regards!

McGiverGim commented:
Hi unitware! One question:
- I have connected by ibus my FS-IA10B to my FS-I6S. Installed the latest stable version of cleanflight. Configured it. All the sticks worked.
- I've added the diode and the resistence, and now it don't wok. I don't know if I need your software modified to let the sticks work or not with the diode.
Can you help me? I'm very newbie in the electronic things ;)

unitware commented:
You need to move from the uart rx to tx. And you need to build a custom firmware from the pull request since it's not merged to any official release.  

Hope you get it working. 

//M

Skickat från min iPhone

> 3 feb. 2017 kl. 20:35 skrev McGiverGim <notifications@github.com>:
> 
> Hi unitware! One question:
> 
> I have connected by ibus my FS-IA10B to my FS-I6S. Installed the latest stable version of cleanflight. Configured it. All the sticks worked.
> I've added the diode and the resistence, and now it don't wok. I don't know if I need your software modified to let the sticks work or not with the diode.
> Can you help me? I'm very newbie in the electronic things ;)
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

McGiverGim commented:
Hi unitware. Thanks for your answer.

I'm trying to build it by myself, but it's impossible.

My first problem is that I cannot clone the git repository. I don't know why but it stucks at 14-15%. I've tried all the solutions found in google but they didn't work.
The next problem will be the build itself, I cannot install bash for windows in my system, I don't know why. I can try with cygwin but first I need to end with the clone.

I don't know if you have a .hex file for Naze32 that you can pass to me :(

I will continue trying...

McGiverGim commented:
Finally I compiled the hex file. I've used another computer and all went ok.

First problem:
- In CLI, when I put the command:

serial 1 1088 1115200 57600 115200 115200

I get a "Parse error" :(

McGiverGim commented:
I've found the problem:

In your instructions (in the commit and in the message):

serial 1 1088 1115200 57600 115200 115200

But there's a 1 extra in 1115200 -> 115200

With this all seems to work :)

unitware commented:
Glad that you managed to build the code and thanks for spotting that typo, I will have a look at hydras comments and fix this in the same time some day. hopefully in the coming week but most probably in the coming month.

Sorry that I did not have a hex handy, but time is lacking at the moment and I was not close to my development rig.

McGiverGim commented:
Thanks to you for your hard work and your help! Waiting for the integration with the master branch... ;)

unitware commented:
@hydra: I have rebased and removed the stuff that I wrongly did merge into my branch while trying to rebase earlier.

I can guess that you will find some changes in the include files a bit unusual. I have tried to clean up the dependencies on include files a bit to make it testable. I strongly believe that a file should include what it needs and not the stuff its include file needs (and their ancestors and so on). It might have worked in the beginning but it makes test fixtures and readability suffer. 

The code has not changed since I flew with it but you never know. Maybe @McGiverGim can give it a spin before merge? (I haven't got any quad in flying state at the moment).

McGiverGim commented:
I will try it as soon I can... I'm building now the dron but I'm waiting for the lipo to arrive. I hope this week they arrive...
My others drones doesn't support cleanflight.
When your changes will be approved, I can modify the configurator to add an option for this useful feature.
Thanks again!

unitware commented:
@McGiverGim It's already in the develoment version, https://github.com/cleanflight/cleanflight-configurator/pull/422

McGiverGim commented:
@unitware thanks again. You were fast 😂

unitware commented:
well its nothing new for this pr, just enable serial rx and ibus telemetry on the same port.

FreshRC commented:
@unitware @McGiverGim 
Hi guys,
I am really new at this, could you please confirm if I understood correctly this .
I have SPRacingF3EVo and FS-iA10b (FS-i6S transmiter) on the way and I would like to do this mod.
So I only need to connect iBus Serial and iBus Sensor with UART TX pin on FC (of course add resistor and diode) and telemetry will work on the same port as iBUS?
Also I need to wait for CF Configurator 1.2.5 to be released to use this feature?

McGiverGim commented:
Yes that is. You can use 1.2.5 developer or use the CLI commands in the actual version.
But this pull request is not included in the latest version of cleanflight, it will be included sometime perhaps 😉. Maybe @hydra can say something about it.

FreshRC commented:
@McGiverGim 
Thank you!
Just one more thing, under "But this pull request is not included in the latest version of cleanflight" you mean that this will not work on CF 1.14.2?
What do I need to do to have it working :-/

McGiverGim commented:
You need to download the code of 1.14.2, merge the changes needed to make this work (that are included into this pull request) and compile all.
If you don't know anything about developing code, maybe I can do it for you, if you are interested.

FreshRC commented:
@McGiverGim 
Hi, I've never liked programming which is weird as I work in IT company :-D
And really don't know anything about developing code.
If it's not too much trouble or waste of your time please I am interested.
As I am new in this RC world and this is my first build it looks to me I will have to start coding if I want to use the full potential of equipment.
So with your code I will only need to flash the FC and it should work (regardless of CF Configurator?)

McGiverGim commented:
Here you have, for the SPRACING3EVO:
https://drive.google.com/file/d/0B6cSJRBCexb2UFNWclNqREUteTA/view?usp=sharing
If someone want to test for NAZE (that's what I use):
https://drive.google.com/file/d/0B6cSJRBCexb2WVRuMXdXVHdiSjA/view?usp=sharing

To activate it, in CLI put this commands (sample for serial 1, I don't know what serial will you use):
>     serial 1 1088 115200 57600 115200 115200
>     feature RX_SERIAL
>     set serialrx_provider = IBUS
>     save

FreshRC commented:
@McGiverGim 
WOW this was fast, thank you very much!
I will use UART2 so I guess in CLI I will add serial 2, right

McGiverGim commented:
Caution! In CLI, serial starts with 0, so for UART2 I think you must use serial 1.

FreshRC commented:
@McGiverGim 
THANK YOU!
Hopefully I will receive all parts today or tomorrow and I will be able to try it out, will let you know how that went ;)

unitware commented:
@McGiverGim did you test fly the PR after I did the cleanup on 5th feb? It's ready for merge as soon as we get some approval of actual flight worthiness :) (My quads are not in a mood to fly at the moment)

McGiverGim commented:
@unitware the lipos finally arrived today :)

This weekend I will try to fly it. I will report my findings... ;)

McGiverGim commented:
@unitware tomorrow I will do more tests, but today I've been flying my new robocat.
I've not had any problem. I've been flying first in acro and after that in horizon, and both modes worked without problem. 
I've been trying to fly in GPS mode but I've problems to get a fix with the GPS. I don't know what can be the problem, but the comunication between the GPS and the Naze is ok, and in gpspassthrough mode I can't connect with the u-center software without problem. Maybe I need to let more time withouth moving the GPS.
Now, I will analize the blackbox log and test again tomorrow, but for me, your patch is ok, but I'm very newbie in the hobbie ;)
Ahhh! I forgot to mention, the telemetry arrive to my FS-i6s withouth problem :)

unitware commented:
@hydra what is going on? This pull request has now been flight tested and working. Why are you closing it without merging it to master?

McGiverGim commented:
All issues and pull request have been closed. Maybe a hacker? Is strange.

unitware commented:
@hydra, please reopen and merge into master_v1.x, i have changed base

McGiverGim commented:
@unitware I don't know what will happen with this PR, but today I've flied 6 lipos. I've tested angle, horizon, acro and airmode. Some flips and rolls. All worked perfectly. 

The problem with the GPS was solved by moving the position of the receiver. Now it locks 8-9 satellites without problem, but flying in GPS hold does not hold anything 🤔 maybe something bad configured.

I will remain with this version by the moment with your changes and probably forever because I don't know where this is going in version 2.0 for F1 users.

Thanks!

unitware commented:
Glad to hear that you are happy with it. I'm still hoping that there will be some efforts on maintaining the master_v1.x branch for us with F1 targets.

unitware commented:
This PR has now risen from the ashes in Betaflight: https://github.com/betaflight/betaflight/pull/2639

dolie commented:
Hello @unitware @McGiverGim , 
I can't make it work :(
I am using 1kohm resistor, 1N4148 diode and a spracingf3.
I entered the commands in CLI but when I check `serial` it is reset to defaults...
Please help 🙏 

```
ER;$M>'&$M>nø
CLI

# serial
serial 0 1 115200 57600 0 115200
serial 1 0 115200 57600 0 115200
serial 2 0 115200 57600 0 115200

# serial 1 1088 115200 57600 115200 115200
serial 1 1088 115200 57600 115200 115200

# feature RX_SERIAL
Enabled RX_SERIAL

# set serialrx_provider = IBUS
serialrx_provider set to IBUS
# save
```

![foto_no_exif](https://user-images.githubusercontent.com/18349922/91031333-b76d3f00-e600-11ea-92b3-058bc2a51ce9.jpg)

dolie commented:
Update : It seems that some features (like USE_TELEMETRY_IBUS) has been removed from boards that are ≤ 256k . I will manually build cleanflight with needed features and try again.

dolie commented:
Final Update : with a custom build (adding especially USE_TELEMETRY_IBUS), I was able to use "on same wire serial and telemetry". The CLI commands did not worked, BUT, enabling serial RX (type IBUS) and IBUS Telemetry together on UART2, it worked. 👍
